packages/auth/src/providers/cognito/utils/signInHelpers.ts

```
@@ -970,3 +991,32 @@ export async function getNewDeviceMetatada(
 		return undefined;
 	}
 }
+
+/**
+ * It will retry the function if the error is a `ResourceNotFoundException` and
+ * will clean the device keys stored in the storage mechanism.
+ *
+ */
+export async function retryOnResourceNotFoundException<
+	F extends (...args: any[]) => any
+>(
+	func: F,
+	args: Parameters<F>,
+	username: string,
+	tokenOrchestrator: AuthTokenOrchestrator
+): Promise<ReturnType<F>> {
+	try {
+		const output = await func(...args);
+		return output;
+	} catch (error) {
+		if (
+			error instanceof AuthError &&
+			error.name === 'ResourceNotFoundException'
```

Should we also check for `error.message === 'Device does not exist.'` to make sure any other errors are not swallowed